Provide strategy for updating Prow cluster (#148) (#747)
kyma-project/test-infra: docs/prow/prow-upgrade.md

## Update process

To update a Prow cluster follow these steps:

1. Follow [this](./prow-installation-on-forks.md) prow-installation-on-forks.md) document for details to set up a Prow cluster.
2. In the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/) project select the commit that is going to be your the one the prow cluster should be upgraded to. E.g. [2c8e0dbb96b4c1a86d42275dfbed5474a6d05def](https://github.com/kubernetes/test-infra/commit/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def).
3. Open both [`starter.yaml`](../../prow/cluster/starter.yaml) in the current project and [`starter.yaml`](https://github.com/kubernetes/test-infra/blob/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def/prow/cluster/starter.yaml) in the Kubernetes project and copy new lables for these containers:
    * gcr.io/k8s-prow/hook
    * gcr.io/k8s-prow/plank
    * gcr.io/k8s-prow/sinker
    * gcr.io/k8s-prow/deck
    * gcr.io/k8s-prow/horologium
    * gcr.io/k8s-prow/tide
4. Copy commit id into comment on top of the file to keep track of the current release used for the deployments (due to prow not having any releases)

> 5. Check with a diff tool for other meaningful changes that need to be taken into account and copied over.

Reviewer: Meaningful changes sound too subjective, please add an example.
```suggestion
5. Use a diff tool to find other meaningful changes and copy the items you need.
```

6. Open both [`config.yaml`](../../prow/config.yaml) in the current project and [`config.yaml`](https://github.com/kubernetes/test-infra/blob/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def/prow/config.yaml) in the Kubernetes project and copy new labels for these containers:
    * gcr.io/k8s-prow/clonerefs
    * gcr.io/k8s-prow/initupload
    * gcr.io/k8s-prow/entrypoint
    * gcr.io/k8s-prow/sidecar

> 7. Use a diff tool to find other meaningful changes and copy the items you need.

Reviewer: Meaningful changes sound too subjective, please add an example.
```suggestion
7. Use a diff tool to find other meaningful changes and copy the items you need.
```

8. Run this command to update Prow deployments:
    ```bash
    kubectl apply -f prow/cluster/starter.yaml
    ```
9. Use the `update-config.sh {file_path}` script to apply the Prow configuration on a cluster. Run the following command:
   ```
   ./update-config.sh ../prow/config.yaml
   ```
10. Make sure that the update was successful. For example, run `kubectl get pods` to check if it doesn't show errors on the updated test cluster and the dashboard is still reachable.
11. Create a pull request.

## Troubleshooting

> 1. In case something goes wrong with the upgrade and pods are not starting anymore, check out the old commit before the update happened and change the deployments back to what they were via:

Reviewer: ```suggestion
If after the update pods don't start, use the previous commit to bring back the previous deployments' version. Follow these steps: 
```

Reviewer: Inbetween what is in line 39 and 40, add the first point of the instruction. You can write the following:

1. Run the following command to update or create a deployment with the latest container version.

# How to update the cluster

> In order to get improved/newer versions of prow we need to upgrade prow. The kubernetes prow instance is getting updated via a shell script that is very limiting in the versions that you can choose from and is probably not good enough for cherrypicking updates we like. In the meantime though we can't not update the cluster and will have to describe a way to do updates. This document aims to provide a description of such an upgrade.

Author: I think I'll have to reword this.

This should've capture the actual way of how the kubernetes team updates their own prow cluster. They use a script called 'bump.sh' that gives back a list of, I think the last five "releases" of their prow images. These could all be from the same day. We're more looking into "We want this feature in our prow" and point this to a specific version, which might be from the time that someone creating the issue on Github, to someone picking up the issue and working on it a couple days out and you won't be able to see that version via the script anymore.

For me it was important to give a "this is how they do it" vs "we should have something similar" statement. This seems to have come across differently than I expected and I'll provide a new paragraph.

## Upgrade process

In general the update process is pretty straight forward and can be rolled back pretty easily.

1. Follow prow-installation-on-forks.md to setup your own prow cluster. Make sure that it's working.

> 2. In the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/) project select the commit that is going to be your the one the prow cluster should be upgraded to. E.g. [2c8e0dbb96b4c1a86d42275dfbed5474a6d05def](https://github.com/kubernetes/test-infra/commit/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def).

Author: Is there an alternative to "select ... choose", this reads weird.

3. Open [starter.yaml (local)](../../prow/cluster/starter.yaml) in current project as well as on the commit [starter.yaml (remote)](https://github.com/kubernetes/test-infra/blob/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def/prow/cluster/starter.yaml)
4. Copy over new labels for the following containers:
    * gcr.io/k8s-prow/hook
    * gcr.io/k8s-prow/plank
    * gcr.io/k8s-prow/sinker
    * gcr.io/k8s-prow/deck
    * gcr.io/k8s-prow/horologium
    * gcr.io/k8s-prow/tide

> 5. Copy commit id into comment on top of the file to keep track of the current release used for the deployments (due to prow not having any releases)

Reviewer: Please clarify what it means that Prow doesn't have any releases.

> 5. Copy commit id into comment on top of the file to keep track of the current release used for the deployments (due to prow not having any releases)

Author: At the moment prow is not available in any versioned release. It just exists as a repository and there is no version 0.1 or 1.0. The containers that make up a prow deployment have unique identifiers that link them but none of them are promised to be stable as far as I know.

I'm not sure this explanation needs to be part of the upgrade process. What's your thought?

> 6. Check with a diff tool for other meaningful changes that need to be taken into account and copied over.

Reviewer: What are the meaningful changes?

> 6. Check with a diff tool for other meaningful changes that need to be taken into account and copied over.

Author: New features introduced by the prow team, that we see useful for our development/testing processes are changes that I'd consider meaningful.

7. Open [config.yaml (local)](../../prow/config.yaml) in current project as well as on the commit [config.yaml (remote)](https://github.com/kubernetes/test-infra/blob/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def/prow/config.yaml)
8. Copy over new labels for the following containers:
    * gcr.io/k8s-prow/clonerefs
    * gcr.io/k8s-prow/initupload
    * gcr.io/k8s-prow/entrypoint
    * gcr.io/k8s-prow/sidecar

> 9. Check with a diff tool for other meaningful changes that need to be taken into account and copied over.

Reviewer: What are the meaningful changes?

10. Update prow deployments
    ```bash
    kubectl apply -f prow/cluster/starter.yaml
    ```
11. Update prow config. Use the `update-config.sh {file_path}` script to apply Prow configuration on a cluster.
   ```
   ./update-config.sh ../prow/config.yaml
   ```
12. Check that everything is working as intended. E.g. that `kubectl get pods` doesn't show errors on your updated test cluster and the dashboard is still reachable.
13. Create PR.

# Worst case rollback

> In case something goes wrong with the upgrade and pods are not starting anymore, check out the old commit before the update happened and change the deployments back to what they were via:

Reviewer: What does the command do?
```bash
kubectl apply -f prow/cluster/starter.yaml
```

> In case something goes wrong with the upgrade and pods are not starting anymore, check out the old commit before the update happened and change the deployments back to what they were via:

Author: On an existing cluster, which is a prerequisite to doing an upgrade, it will update the deployment with the newer versions of existing containers and if there's a new container described it will create this deployment.

# Prow cluster update

In order to get improved/newer versions of prow we need to upgrade prow. The kubernetes teams' prow instance is getting updated via a shell script. This shell script offers a small list of the last pushed container tags. This could reflect only containers of a single day, due to how many times the kubernetes team is actually running their pipeline. This limits us in the versions we can choose from, as we usually don't go from finding out about a new version, to deploying it, within one day. To cherrypick updates, we will need a different process.

In the meantime though we can't not update the cluster and will have to describe a way to do updates. This document aims to provide a description of such an upgrade.

## Update process

To update a Prow cluster follow these steps:

> 1. Follow [this](./prow-installation-on-forks.md) document for details to set up a Prow cluster.

Reviewer 2: I think we should add a step where we need to check the announcements, especially the breaking changes, from here: https://github.com/kubernetes/test-infra/blob/master/prow/ANNOUNCEMENTS.md

Author: added this in the first paragraph as a general thing.

2. Go to the [`kubernetes/test-infra`](https://github.com/kubernetes/test-infra/) project and select a commit with the desired upgrade for the Prow cluster. For example, use [2c8e0dbb96b4c1a86d42275dfbed5474a6d05def](https://github.com/kubernetes/test-infra/commit/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def).
3. Open both [`starter.yaml`](../../prow/cluster/starter.yaml) in the current project and [`starter.yaml`](https://github.com/kubernetes/test-infra/blob/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def/prow/cluster/starter.yaml) in the Kubernetes project and copy new lables for these containers:
    * gcr.io/k8s-prow/hook
    * gcr.io/k8s-prow/plank
    * gcr.io/k8s-prow/sinker
    * gcr.io/k8s-prow/deck
    * gcr.io/k8s-prow/horologium
    * gcr.io/k8s-prow/tide
4. Copy the commit ID into a comment at the top of the file to keep track of the current release used for the deployments.
5. Check with a diff tool for other meaningful changes (e.g. stability upgrade in a prow component) that need to be taken into account and copied over.

> 6. Open both [`config.yaml`](../../prow/config.yaml) in the current project and [`config.yaml`](https://github.com/kubernetes/test-infra/blob/2c8e0dbb96b4c1a86d42275dfbed5474a6d05def/prow/config.yaml) in the Kubernetes project and copy new labels for these containers:

Reviewer 2: I'd use `tag` instead of `label` to make it more clear.
